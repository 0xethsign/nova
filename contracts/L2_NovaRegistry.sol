// SPDX-License-Identifier: AGPL-3.0-only
pragma solidity 0.7.6;
pragma abicoder v2;

import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
import "@openzeppelin/contracts/utils/ReentrancyGuard.sol";

contract L2_NovaRegistry is ReentrancyGuard {
    using SafeERC20 for IERC20;

    IERC20 immutable ETH = IERC20(0x4200000000000000000000000000000000000006);
    uint256 immutable MIN_CANCEL_SECONDS = 300;

    /// @notice Emitted when `cancel` is called.
    /// @param timestamp When the cancel will set into effect and the creator will be able to withdraw.
    event Cancel(bytes32 indexed execHash, uint256 timestamp);

    /// @notice Emitted when `withdraw` is called.
    event Withdraw(bytes32 indexed execHash);

    /// @notice Emitted when `requestExec` is called.
    event Request(bytes32 indexed execHash, address indexed task);

    /// @notice Emitted when `bumpGas` is called.
    /// @param newExecHash The execHash of the resubmitted request (copy of its uncle with an updated gasPrice).
    /// @param timestamp When the uncled request (`execHash`) will have its tokens transfered to the resubmitted request (`newExecHash`).
    event BumpGas(
        bytes32 indexed execHash,
        bytes32 indexed newExecHash,
        uint256 timestamp
    );

    /// @notice Emitted when `execCompleted` is called.
    event ExecCompleted(
        bytes32 indexed execHash,
        address indexed executor,
        address indexed rewardRecipient,
        uint256 gasUsed,
        bool reverted
    );

    struct InputToken {
        IERC20 l2Token;
        address l1Token;
        uint256 amount;
    }

    struct Bounty {
        IERC20 token;
        uint256 amount;
    }

    /// @dev The most recent nonce assigned to an execution request.
    uint72 private systemNonce;

    /// @dev Maps execHashes to the creator of each request.
    mapping(bytes32 => address) private requestCreators;
    /// @dev Maps execHashes to the address of the task associated with the request.
    mapping(bytes32 => address) private requestTasks;
    /// @dev Maps execHashes to the calldata associated with the request.
    mapping(bytes32 => bytes) private requestCalldatas;
    /// @dev Maps execHashes to the gas limit a bot should use to execute the request.
    mapping(bytes32 => uint32) private requestGasLimits;
    /// @dev Maps execHashes to the gas price a bot must use to execute the request.
    mapping(bytes32 => uint256) private requestGasPrices;
    /// @dev Maps execHashes to the 'bounty' tokens a bot will recieve for executing the request.
    mapping(bytes32 => Bounty[]) private requestBounties;
    /// @dev Maps execHashes to the input tokens a bot must have to execute the request.
    mapping(bytes32 => InputToken[]) private requestInputTokens;

    /// @dev Maps execHashes to a timestamp representing when the request has/will have its tokens removed (via bumpGas/withdraw/execCompleted).
    /// @dev If the request has had its tokens removed via withdraw or execCompleted it will have a timestamp of 1.
    /// @dev If the request will have its tokens removed in the future (via bumpGas) it will be a standard timestamp.
    mapping(bytes32 => uint256) private requestTokenRemovalTimestamps;

    /// @dev Maps execHashes to a timestamp representing when the request is fully canceled and the creator can withdraw their bounties/inputs.
    /// @dev Bots should not attempt to execute a request if the current time has passed its cancel timestamp.
    mapping(bytes32 => uint256) private requestCancelTimestamps;

    /// @dev Maps execHashes which represent resubmitted requests (via bumpGas) to their corresponding "uncled" request's execHash.
    /// @dev An uncled request is a request that has had its tokens removed via `bumpGas` in favor of a resubmitted request generated in the transaction.
    mapping(bytes32 => bytes32) private uncles;

    /// @notice Returns all relevant data about a specific request.
    function getRequestData(bytes32 execHash)
        external
        view
        returns (
            // General request data:
            address task,
            bytes memory l1calldata,
            uint32 gasLimit,
            uint256 gasPrice,
            InputToken[] memory inputTokens,
            Bounty[] memory bounties,
            // Other data:
            address creator,
            bytes32 uncle,
            // Can be fetched via `isExecutable`:
            bool executable,
            uint256 changeTimestamp
        )
    {
        task = requestTasks[execHash];
        l1calldata = requestCalldatas[execHash];
        gasLimit = requestGasLimits[execHash];
        gasPrice = requestGasPrices[execHash];
        inputTokens = requestInputTokens[execHash];
        bounties = requestBounties[execHash];
        creator = requestCreators[execHash];
        uncle = uncles[execHash];

        (executable, changeTimestamp) = isExecutable(execHash);
    }

    /// @param task The address of the "task" contract on L1 a bot should call with `calldata`.
    /// @param l1calldata The abi encoded calldata a bot should call the `task` with on L1.
    /// @param gasLimit The gas limit a bot should use on L1.
    /// @param gasPrice The gas price a bot should use on L1.
    /// @param inputTokens An array of token amounts that a bot will need on L1 to execute the request (`l1Token`s) along with the equivalent tokens that will be returned on L2 (`l2Token`s). `inputTokens` will not be awarded if the `task` reverts on L1.
    /// @param bounties An array of tokens that will be awarded to the bot who executes the request. Only 50% of the bounty will be paid to the bot if the `task` reverts on L1.
    function requestExec(
        address task,
        bytes calldata l1calldata,
        uint32 gasLimit,
        uint256 gasPrice,
        InputToken[] calldata inputTokens,
        Bounty[] calldata bounties
    ) public nonReentrant returns (bytes32 execHash) {
        systemNonce += 1;
        execHash = keccak256(
            abi.encodePacked(systemNonce, task, l1calldata, gasPrice)
        );

        requestTasks[execHash] = task;
        requestCalldatas[execHash] = l1calldata;
        requestGasLimits[execHash] = gasLimit;
        requestGasPrices[execHash] = gasPrice;
        requestCreators[execHash] = msg.sender;

        // Transfer in ETH to pay for max gas usage.
        ETH.safeTransferFrom(msg.sender, address(this), gasPrice * gasLimit);

        // Transfer input tokens in that the msg.sender has approved.
        for (uint256 i = 0; i < inputTokens.length; i++) {
            inputTokens[i].l2Token.safeTransferFrom(
                msg.sender,
                address(this),
                inputTokens[i].amount
            );

            // Copy over this index to the requestInputTokens mapping (we can't just put a calldata/memory array directly into storage so we have to go index by index).
            requestInputTokens[execHash][i] = inputTokens[i];
        }

        // Transfer bounties in that the msg.sender has approved.
        for (uint256 i = 0; i < bounties.length; i++) {
            bounties[i].token.safeTransferFrom(
                msg.sender,
                address(this),
                bounties[i].amount
            );

            // Copy over this index to the requestBounties mapping (we can't just put a calldata/memory array directly into storage so we have to go index by index).
            requestBounties[execHash][i] = bounties[i];
        }

        emit Request(execHash, task);
    }

    /// @notice Calls `requestExec` with all relevant parameters along with calling `cancel` with the `autoCancelDelay` argument.
    /// @dev See `requestExec` and `cancel` for more information.
    function requestExecWithTimeout(
        address task,
        bytes calldata l1calldata,
        uint32 gasLimit,
        uint256 gasPrice,
        InputToken[] calldata inputTokens,
        Bounty[] calldata bounties,
        uint256 autoCancelDelay
    ) external returns (bytes32 execHash) {
        execHash = requestExec(
            task,
            l1calldata,
            gasLimit,
            gasPrice,
            inputTokens,
            bounties
        );

        cancel(execHash, autoCancelDelay);
    }

    /// @notice Cancels a request with a delay. Once the delay has passed anyone may call `withdraw` on behalf of the user to recieve their bounties/input tokens back.
    /// @notice msg.sender must be the creator of the request associated with the `execHash`.
    /// @param execHash The unique hash of the request to cancel.
    /// @param withdrawDelaySeconds The delay in seconds until the creator can withdraw their tokens. Must be greater than or equal to `MIN_CANCEL_SECONDS`.
    function cancel(bytes32 execHash, uint256 withdrawDelaySeconds) public {
        (bool tokensRemoved, ) = areTokensRemoved(execHash);
        require(!tokensRemoved, "TOKENS_REMOVED");
        require(requestCancelTimestamps[execHash] == 0, "ALREADY_CANCELED");
        require(requestCreators[execHash] == msg.sender, "NOT_CREATOR");
        require(withdrawDelaySeconds >= MIN_CANCEL_SECONDS, "DELAY_TOO_SMALL");

        // Set the delay timestamp to int(current timestamp + the delay)
        uint256 timestamp = block.timestamp + withdrawDelaySeconds;
        requestCancelTimestamps[execHash] = timestamp;

        emit Cancel(execHash, timestamp);
    }

    /// @notice Withdraws tokens (input/gas/bounties) from a canceled task.
    /// @notice The creator of the request associated with `execHash` must call `cancel` and wait the `withdrawDelaySeconds` they specified before calling `withdraw`.
    /// @notice Anyone may call this method on behalf of another user but the tokens will still go the creator of the request associated with the `execHash`.
    /// @param execHash The unique hash of the request to withdraw from.
    function withdraw(bytes32 execHash) external nonReentrant {
        (bool tokensRemoved, ) = areTokensRemoved(execHash);
        require(!tokensRemoved, "TOKENS_REMOVED");
        (bool canceled, ) = isCanceled(execHash);
        require(canceled, "NOT_CANCELED");

        address creator = requestCreators[execHash];
        InputToken[] memory inputTokens = requestInputTokens[execHash];
        Bounty[] memory bounties = requestBounties[execHash];

        // Store that the request has had its tokens removed.
        requestTokenRemovalTimestamps[execHash] = 1;

        // Transfer the ETH which would have been used for gas back to the creator.
        ETH.transfer(
            creator,
            requestGasPrices[execHash] * requestGasLimits[execHash]
        );

        // Transfer input tokens back to the creator.
        for (uint256 i = 0; i < inputTokens.length; i++) {
            inputTokens[i].l2Token.safeTransfer(creator, inputTokens[i].amount);
        }
        // Transfer bounties back to the creator.
        for (uint256 i = 0; i < bounties.length; i++) {
            bounties[i].token.safeTransfer(creator, bounties[i].amount);
        }

        emit Withdraw(execHash);
    }

    /// @notice
    function bumpGas(bytes32 execHash, uint256 gasPrice)
        external
        returns (bytes32 newExecHash)
    {
        (bool executable, ) = isExecutable(execHash);
        require(executable, "NOT_EXECUTABLE");
        uint256 previousGasPrice = requestGasPrices[execHash];
        require(requestCreators[execHash] == msg.sender, "NOT_CREATOR");
        require(gasPrice > previousGasPrice, "LESS_THAN_PREVIOUS_GAS_PRICE");

        systemNonce += 1;
        newExecHash = keccak256(
            abi.encodePacked(
                systemNonce,
                requestTasks[execHash],
                requestCalldatas[execHash],
                gasPrice
            )
        );

        uint32 gasLimit = requestGasLimits[execHash];

        requestTasks[newExecHash] = requestTasks[execHash];
        requestCalldatas[newExecHash] = requestCalldatas[execHash];
        requestGasLimits[newExecHash] = gasLimit;
        requestGasPrices[newExecHash] = gasPrice;
        requestCreators[newExecHash] = msg.sender;

        // Map the resubmitted request to its uncle.
        uncles[newExecHash] = execHash;

        // Set the uncled request to expire in MIN_CANCEL_SECONDS.
        uint256 switchTimestamp = MIN_CANCEL_SECONDS + block.timestamp;
        requestTokenRemovalTimestamps[execHash] = switchTimestamp;

        // Transfer in additional ETH to pay for the new gas limit.
        ETH.safeTransferFrom(
            msg.sender,
            address(this),
            (gasPrice - previousGasPrice) * gasLimit
        );

        emit BumpGas(execHash, newExecHash, switchTimestamp);
    }

    function execCompleted(
        bytes32 execHash,
        address executor,
        address rewardRecipient,
        uint256 gasUsed,
        bool reverted
    ) external nonReentrant {
        (bool executable, ) = isExecutable(execHash);
        require(executable, "NOT_EXECUTABLE");

        // Store that the request has had its tokens removed.
        requestTokenRemovalTimestamps[execHash] = 1;

        InputToken[] memory inputTokens = requestInputTokens[execHash];
        Bounty[] memory bounties = requestBounties[execHash];

        // Transfer the ETH used for gas to the rewardRecipient.
        ETH.transfer(rewardRecipient, requestGasPrices[execHash] * gasUsed);

        // Only transfer input tokens if the request didn't revert.
        if (!reverted) {
            // Transfer input tokens to the rewardRecipient.
            for (uint256 i = 0; i < inputTokens.length; i++) {
                inputTokens[i].l2Token.safeTransfer(
                    rewardRecipient,
                    inputTokens[i].amount
                );
            }

            // Transfer full bounty back to the rewardRecipient.
            for (uint256 i = 0; i < bounties.length; i++) {
                bounties[i].token.safeTransfer(
                    rewardRecipient,
                    bounties[i].amount
                );
            }
        } else {
            address creator = requestCreators[execHash];

            // Transfer input tokens back to the creator.
            for (uint256 i = 0; i < inputTokens.length; i++) {
                inputTokens[i].l2Token.safeTransfer(
                    creator,
                    inputTokens[i].amount
                );
            }

            // Transfer 70% of the bounty to the rewardRecipient and 30% back to the creator.
            for (uint256 i = 0; i < bounties.length; i++) {
                IERC20 token = bounties[i].token;

                token.safeTransfer(
                    rewardRecipient,
                    // 70% goes to the rewardRecipient
                    (bounties[i].amount * 3) / 10
                );

                token.safeTransfer(
                    creator,
                    // 30% goes to the creator
                    (bounties[i].amount * 3) / 10
                );
            }
        }

        emit ExecCompleted(
            execHash,
            executor,
            rewardRecipient,
            gasUsed,
            reverted
        );
    }

    /// @notice Checks if the request is executable along with a timestamp of when that may change.
    /// @return executable A boolean indicating if the request is executable.
    /// @return changeTimestamp A timestamp indicating when the request might switch from being executable to unexecutable (or vice-versa). Will be 0 if there is no change expected. It will be a timestamp if the request will be enabled soon (it's a resubmitted version of an uncled request) or the request is being canceled soon.
    function isExecutable(bytes32 execHash)
        public
        view
        returns (bool executable, uint256 changeTimestamp)
    {
        if (requestCreators[execHash] == address(0)) {
            // This isn't a valid execHash!
            executable = false;
            changeTimestamp = 0;
        } else {
            (bool tokensRemoved, uint256 tokensRemovedChangeTimestamp) =
                areTokensRemoved(execHash);
            (bool canceled, uint256 canceledChangeTimestamp) =
                isCanceled(execHash);

            executable = !tokensRemoved && !canceled;

            // One or both of these values will be 0 so we can just add them.
            changeTimestamp =
                canceledChangeTimestamp +
                tokensRemovedChangeTimestamp;
        }
    }

    /// @notice Checks if the request is currently canceled along with a timestamp of when it may be canceled.
    /// @return tokensRemoved A boolean indicating if the request has been canceled.
    /// @return changeTimestamp A timestamp indicating when the request might have its tokens removed or added. Will be 0 if there is no removal/addition expected. It will be a timestamp if the request will have its tokens added soon (it's a resubmitted version of an uncled request).
    function areTokensRemoved(bytes32 execHash)
        public
        view
        returns (bool tokensRemoved, uint256 changeTimestamp)
    {
        uint256 removalTimestamp = requestTokenRemovalTimestamps[execHash];

        if (removalTimestamp == 0) {
            bytes32 uncle = uncles[execHash];

            // Check if this request is a resubmitted version of an uncled request.
            if (uncle.length == 0) {
                // This is a normal request, so we know tokens have/will not been removed.
                tokensRemoved = false;
                changeTimestamp = 0;
            } else {
                // This is a resubmitted version of a uncled request, so we have to check if the uncle has had its tokens removed,
                // if so, this request has its tokens.
                uint256 uncleDeathTimestamp =
                    requestTokenRemovalTimestamps[uncle];

                if (uncleDeathTimestamp == 1) {
                    // The uncle request has had its tokens removed early.
                    tokensRemoved = true;
                    changeTimestamp = 0;
                } else {
                    // The uncled request may still be waiting for its tokens to be removed.
                    tokensRemoved = block.timestamp < uncleDeathTimestamp; // Tokens are removed for a resubmitted request if the uncled request has not had its tokens removed yet.
                    changeTimestamp = tokensRemoved
                        ? uncleDeathTimestamp // Return a timestamp if the request is still waiting to have tokens added.
                        : 0;
                }
            }
        } else {
            // Tokens have/will be removed.
            tokensRemoved = block.timestamp >= removalTimestamp; // Tokens are removed if the current timestamp is greater than the removal timestamp.
            changeTimestamp = tokensRemoved ? 0 : removalTimestamp; // Return a timestamp if the tokens have not been removed yet.
        }
    }

    /// @notice Checks if the request is currently canceled along with a timestamp of when it may be canceled.
    /// @return canceled A boolean indicating if the request has been canceled.
    /// @return changeTimestamp A timestamp indicating when the request might be canceled. Will be 0 if there is no cancel expected. It will be a timestamp if a cancel has been requested.
    function isCanceled(bytes32 execHash)
        public
        view
        returns (bool canceled, uint256 changeTimestamp)
    {
        uint256 cancelTimestamp = requestCancelTimestamps[execHash];

        if (cancelTimestamp == 0) {
            // There has been no cancel attempt.
            canceled = false;
            changeTimestamp = 0;
        } else {
            // There has been a cancel attempt.
            canceled = block.timestamp >= cancelTimestamp;
            changeTimestamp = canceled ? 0 : cancelTimestamp;
        }
    }
}
