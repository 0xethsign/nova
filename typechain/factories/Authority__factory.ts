/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import { Provider } from "@ethersproject/providers";
import type { Authority, AuthorityInterface } from "../Authority";

const _abi = [
  {
    inputs: [
      {
        internalType: "address",
        name: "src",
        type: "address",
      },
      {
        internalType: "address",
        name: "dst",
        type: "address",
      },
      {
        internalType: "bytes4",
        name: "sig",
        type: "bytes4",
      },
    ],
    name: "canCall",
    outputs: [
      {
        internalType: "bool",
        name: "",
        type: "bool",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
];

export class Authority__factory {
  static readonly abi = _abi;
  static createInterface(): AuthorityInterface {
    return new utils.Interface(_abi) as AuthorityInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): Authority {
    return new Contract(address, _abi, signerOrProvider) as Authority;
  }
}
