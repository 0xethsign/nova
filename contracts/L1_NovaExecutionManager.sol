// @unsupported: ovm
// SPDX-License-Identifier: AGPL-3.0-only
pragma solidity 0.7.6;
pragma abicoder v2;

import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
import "@eth-optimism/contracts/libraries/bridge/OVM_CrossDomainEnabled.sol";
import "./L2_NovaRegistry.sol";

contract L1_NovaExecutionManager is OVM_CrossDomainEnabled {
    using SafeERC20 for IERC20;

    /// @dev The revert message text used to cause a hard revert.
    string private constant HARD_REVERT_TEXT = "__NOVA__HARD__REVERT__";
    /// @dev The hash of the hard revert message.
    bytes32 private constant HARD_REVERT_HASH = keccak256(abi.encodeWithSignature("Error(string)", HARD_REVERT_TEXT));

    /// @dev The bytes length of an abi encoded execCompleted call.
    uint256 public constant execCompletedMessageBytesLength = 164;

    /// @dev The address of the L2_NovaRegistry to send cross domain messages to.
    address public immutable L2_NovaRegistryAddress;

    /// @notice Maps execHashes to a boolean indicating whether the corresponding request has already been executed.
    mapping(bytes32 => bool) public executed;

    /// @dev The execHash computed from the currently executing call to `exec`.
    /// @dev This will be reset after every execution.
    bytes32 public currentExecHash;
    /// @dev The address who called `exec`/`execWithRecipient`.
    /// @dev This will not be reset to address(0) after each execution completes.
    address public currentExecutor;
    /// @dev The address of the strategy that is currenlty being called.
    /// @dev This will not be reset to address(0) after each execution completes.
    address public currentlyExecutingStrategy;

    constructor(address _L2_NovaRegistryAddress, address _messenger) OVM_CrossDomainEnabled(_messenger) {
        L2_NovaRegistryAddress = _L2_NovaRegistryAddress;
    }

    function execWithRecipient(
        uint72 nonce,
        address strategy,
        bytes calldata l1calldata,
        uint32 xDomainMessageGasLimit,
        address l2Recipient
    ) public {
        uint256 startGas = gasleft();

        // Compute the execHash.
        bytes32 execHash = keccak256(abi.encodePacked(nonce, strategy, l1calldata, tx.gasprice));

        // Prevent double executing.
        require(!executed[execHash], "ALREADY_EXECUTED");

        // Initialize execution context.
        currentExecHash = execHash;
        currentExecutor = msg.sender;
        currentlyExecutingStrategy = strategy;

        // Call the strategy.
        (bool success, bytes memory returnData) = strategy.call(l1calldata);

        // Revert if the strategy hard reverted.
        require(success || !isHardRevert(returnData), "HARD_REVERT");

        // Mark the request as executed.
        executed[execHash] = true;

        // Reset execution context.
        // We reset only one of the execution context variables because it will cost us less gas to use a previously set storage slot on all future runs.
        delete currentExecHash;

        // Figure out how much gas this xDomain message is going to cost us.
        uint256 xDomainMessageGas =
            // ((estimated cost per calldata char) * (bytes length for an encoded call to execCompleted)) + ((cross domain gas limit) / (enqueue gas burn)) + (sendMessage overhead)
            (48 * execCompletedMessageBytesLength) + (xDomainMessageGasLimit / 32) + 74000;

        // Figure out how much gas this call will take up in total: (Constant function call gas) + (Gas diff after calls) + (the amount of gas that will be burned via enqueue + storage/other message overhead)
        uint256 gasUsed = 21396 + (startGas - gasleft()) + xDomainMessageGas;

        // Send message to unlock the bounty on L2.
        sendCrossDomainMessage(
            L2_NovaRegistryAddress,
            abi.encodeWithSelector(
                L2_NovaRegistry(L2_NovaRegistryAddress).execCompleted.selector,
                execHash,
                msg.sender,
                l2Recipient,
                gasUsed,
                !success
            ),
            xDomainMessageGasLimit
        );
    }

    function exec(
        uint72 nonce,
        address strategy,
        bytes calldata l1calldata,
        uint32 xDomainMessageGasLimit
    ) external {
        execWithRecipient(nonce, strategy, l1calldata, xDomainMessageGasLimit, msg.sender);
    }

    function hardRevert() external pure {
        // Call revert with the hard revert text.
        revert(HARD_REVERT_TEXT);
    }

    function transferFromBot(address token, uint256 amount) external {
        // Only the currently executing strategy is allowed to call this method.
        // Must check that the execHash is not empty first to make sure that there is an execution in-progress.
        require(currentExecHash.length > 0 && msg.sender == currentlyExecutingStrategy, "NOT_CURRENTLY_EXECUTING");

        // Transfer the token from the calling bot the currently executing strategy (msg.sender is enforced to be the currentlyExecutingStrategy above).
        IERC20(token).safeTransferFrom(currentExecutor, msg.sender, amount);
    }

    function isHardRevert(bytes memory returnData) private pure returns (bool) {
        // We know the reverting with the HARD_REVERT_TEXT results in returnData with a length of 100.
        if (returnData.length != 100) return false;

        // Check if the revert data matches the HARD_REVERT_HASH.
        return keccak256(returnData) == HARD_REVERT_HASH;
    }
}
