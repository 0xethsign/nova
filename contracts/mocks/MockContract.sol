// @unsupported: ovm
// SPDX-License-Identifier: AGPL-3.0-only
pragma solidity 0.7.6;

import "../L1_NovaExecutionManager.sol";

contract MockContract {
    function thisFunctionWillNotRevert() external {}

    function thisFunctionWillRevert() external {
        revert("Not a hard revert!");
    }

    function thisFunctionWillHardRevert() external {
        L1_NovaExecutionManager(msg.sender).hardRevert();
    }
}
