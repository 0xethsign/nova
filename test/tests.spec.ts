import chai from "chai";
import chaiAsPromised from "chai-as-promised";

import { getContractFactory } from "@eth-optimism/contracts";

import { Watcher } from "@eth-optimism/core-utils";

import {
  createFactory,
  createTestWallet,
  wait,
  waitForL1ToL2Tx,
} from "../utils/testUtils";
import { ethers, network } from "hardhat";

import {
  L2NovaRegistry__factory,
  L2NovaRegistry,
  ERC20,
  MockOVMETH__factory,
  MockCrossDomainMessenger__factory,
} from "../typechain";
import { L1NovaExecutionManager } from "../typechain/L1NovaExecutionManager";
import { L1NovaExecutionManager__factory } from "../typechain/factories/L1NovaExecutionManager__factory";
import { MockContract__factory } from "../typechain/factories/MockContract__factory";
import { Wallet } from "@ethersproject/wallet";
import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
import { MockContract } from "../typechain/MockContract";

chai.use(chaiAsPromised);
chai.should();

function computeExecHash({
  nonce,
  strategy,
  calldata,
  gasPrice,
}: {
  nonce: number;
  strategy: string;
  calldata: string;
  gasPrice: number;
}) {
  return ethers.utils.solidityKeccak256(
    ["uint72", "address", "bytes", "uint256"],
    [nonce, strategy, calldata, gasPrice]
  );
}

describe("Nova", function () {
  let l1Wallet: Wallet | SignerWithAddress;
  let l2Wallet: Wallet | SignerWithAddress;

  // Cross layer watcher:
  let watcher: Watcher;

  // Nova specific contracts:
  let l1_NovaExecutionManager: L1NovaExecutionManager;
  let l2_NovaRegistry: L2NovaRegistry;

  // OVM Contracts:
  let OVM_ETH: ERC20;

  before(async () => {
    l1Wallet = network.ovm
      ? createTestWallet("http://localhost:9545")
      : (await ethers.getSigners())[0];
    l2Wallet = network.ovm
      ? createTestWallet("http://localhost:8545")
      : (await ethers.getSigners())[0];

    // Get us some OVM_ETH:
    if (network.ovm) {
      OVM_ETH = getContractFactory("OVM_ETH")
        .connect(l2Wallet)
        .attach("0x4200000000000000000000000000000000000006") as ERC20;

      const OVM_L1ETHGateway = getContractFactory("OVM_L1ETHGateway")
        .connect(l1Wallet)
        .attach("0x998abeb3e57409262ae5b751f60747921b33613e");

      // Create a watcher:
      watcher = new Watcher({
        l1: {
          provider: l1Wallet.provider,
          messengerAddress: "0x59b670e9fA9D0A427751Af201D676719a970857b",
        },
        l2: {
          provider: l2Wallet.provider,
          messengerAddress: "0x4200000000000000000000000000000000000007",
        },
      });

      // Deposit ETH to our L2 wallet:
      await waitForL1ToL2Tx(
        OVM_L1ETHGateway.connect(l1Wallet).depositTo(l2Wallet.address, {
          value: ethers.utils.parseEther("5"),
        }),
        watcher
      );
    } else {
      OVM_ETH = await createFactory<MockOVMETH__factory>(
        false,
        "MockOVMETH",
        "mocks/"
      )
        .connect(l2Wallet)
        .deploy();

      const MockCrossDomainMessenger = await createFactory<MockCrossDomainMessenger__factory>(
        false,
        "MockCrossDomainMessenger",
        "mocks/"
      )
        .connect(l1Wallet)
        .deploy();

      // We don't actually use this watcher it's just a way to pass down the mssenger address.
      watcher = new Watcher({
        l1: {
          provider: l1Wallet.provider,
          messengerAddress: MockCrossDomainMessenger.address,
        },
        l2: {
          provider: l2Wallet.provider,
          messengerAddress: MockCrossDomainMessenger.address,
        },
      });
    }

    // Deploy registry on L2.
    l2_NovaRegistry = await createFactory<L2NovaRegistry__factory>(
      network.ovm,
      "L2_NovaRegistry"
    )
      .connect(l2Wallet)
      .deploy(OVM_ETH.address, watcher.l2.messengerAddress);

    // Deploy execution manager on L1.
    l1_NovaExecutionManager = await createFactory<L1NovaExecutionManager__factory>(
      false,
      "L1_NovaExecutionManager"
    )
      .connect(l1Wallet)
      .deploy(l2_NovaRegistry.address, watcher.l1.messengerAddress);

    // Tell the registry about the execution manager's L1 address.
    await wait(
      l2_NovaRegistry.connectExecutionManager(l1_NovaExecutionManager.address)
    );
  });

  describe("L2_NovaRegistry", function () {
    describe("requestExec", function () {
      it("should allow a valid request", async function () {
        // Approve 100 wei as gas for the first request.
        await wait(OVM_ETH.approve(l2_NovaRegistry.address, 100));

        // This will not revert because we have approved just enough wei.
        await l2_NovaRegistry
          .connect(l2Wallet)
          .requestExec(
            "0x0000000000000000000000000000000000000000",
            "0x00",
            10,
            10,
            [],
            []
          )
          .should.emit(l2_NovaRegistry, "Request")
          .withArgs(
            computeExecHash({
              nonce: 1,
              strategy: "0x0000000000000000000000000000000000000000",
              calldata: "0x00",
              gasPrice: 10,
            }),
            "0x0000000000000000000000000000000000000000"
          );

        // Ensure the registry transferred in the ETH.
        await OVM_ETH.balanceOf(
          l2_NovaRegistry.address
        ).should.eventually.equal(100);
      });

      it("should revert if not enough wei is approved to pay for gas", async function () {
        // This should revert as the 100 wei was already taken by the previous request.
        await l2_NovaRegistry
          .connect(l2Wallet)
          .requestExec(
            "0x0000000000000000000000000000000000000000",
            "0x00",
            10,
            10,
            [],
            []
          ).should.be.reverted;
      });
    });
  });

  describe("L1_NovaExecutionManager", function () {
    it("execCompletedMessageBytesLength should be correct", async function () {
      const execCompletedMessageBytesLength = (
        await l1_NovaExecutionManager.execCompletedMessageBytesLength()
      ).toNumber();

      // Encode a call to execCompleted
      const bytes = l2_NovaRegistry.interface.encodeFunctionData(
        "execCompleted",
        [
          ethers.utils.keccak256("0x00"),
          "0x0000000000000000000000000000000000000000",
          "0x0000000000000000000000000000000000000000",
          0,
          false,
        ]
      );

      // Length of the encoded data in a bytes array.
      const bytesLength = (bytes.length - 2) / 2;

      // The `execCompletedMessageBytesLength` variable should equal the proper bytesLength we just computed.
      execCompletedMessageBytesLength.should.equal(bytesLength);
    });

    describe("exec", function () {
      let mockContract: MockContract;

      before(async function () {
        mockContract = await createFactory<MockContract__factory>(
          false,
          "MockContract",
          "mocks/"
        )
          .connect(l1Wallet)
          .deploy();
      });

      it("should properly execute a request that soft reverts", async function () {
        await l1_NovaExecutionManager.exec(
          // Nonce
          0,
          // Strategy
          mockContract.address,
          // Calldata
          mockContract.interface.encodeFunctionData("thisFunctionWillRevert"),
          // xDomain Gas Limit
          100000
        ).should.not.be.reverted;
      });

      it("shouldn't allow double executing", async function () {
        await l1_NovaExecutionManager.exec(
          // Nonce
          0,
          // Strategy
          mockContract.address,
          // Calldata
          mockContract.interface.encodeFunctionData("thisFunctionWillRevert"),
          // xDomain Gas Limit
          100000
        ).should.be.reverted;
      });

      it("should properly handle a hard revert", async function () {
        await l1_NovaExecutionManager.exec(
          // Nonce
          0,
          // Strategy
          mockContract.address,
          // Calldata
          mockContract.interface.encodeFunctionData(
            "thisFunctionWillHardRevert"
          ),
          // xDomain Gas Limit
          100000
        ).should.be.reverted;
      });
    });
  });
});
