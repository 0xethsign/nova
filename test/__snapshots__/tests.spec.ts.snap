// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nova L1_NovaExecutionManager exec should properly execute a request 1`] = `137323`;

exports[`Nova L1_NovaExecutionManager exec should properly execute a request that soft reverts 1`] = `152540`;

exports[`Nova L2_NovaRegistry requestExec should allow a valid request 1`] = `191068`;

exports[`Nova L2_NovaRegistry should allow connecting to an execution manager once 1`] = `44039`;
